src/main/scala/is/hail/utils/SortedDistinctRowIterator.scala

```diff
@@ -0,0 +1,45 @@
+package is.hail.utils
+
+import is.hail.expr._
+import is.hail.annotations._
+import is.hail.sparkextras._
+
+object SortedDistinctRowIterator {
+  def transformer(ort: OrderedRDD2Type): Iterator[RegionValue] => SortedDistinctRowIterator =
+    new SortedDistinctRowIterator(ort, _)
+}
+
+class SortedDistinctRowIterator(ort: OrderedRDD2Type, it: Iterator[RegionValue]) extends Iterator[RegionValue] {
+  private val bit = it.buffered
+  private val wrv: WritableRegionValue = WritableRegionValue(ort.rowType)
+  private var haveNext = false
+  private var stale = false
```

Comment:
OK, I'm going to start sounding like a broken record.  What's the invariant between `stale`, `haveNext` and `wrv`?  In particular, I'm having trouble understanding how `stale` is different than `!haveNext`.  Also, is `haveNext` just `hasNext` with a different spelling?

Reply:
In Scala, I can't name a private field and a no-parentheses method the same name, apparently. It's also convenient that "have" can be used for the first-person and "has" is third-person.

The iterator is `stale` if the contents of `haveNext` and `wrv` do not necessarily represent the state of the iterator. If the iterator is not `stale`, then `hasNext` = `haveNext` and `next` should return `wrv.value`.

After object initialization, the iterator is always not `stale`. `haveNext` indicates if there is at least one more value. `wrv.value` is the next value, if `haveNext` is true.

After calling `next`, the iterator is `stale`. We don't yet know if `bit` contains another key-distinct value. If `hasNext` or `next` is called, then `SortedDistinctRowIterator` uses `advance` to determine if there is another value, whose key is distinct from the previously returned value.

---

I think I can remove `stale` if I allocate a second `RegionValue` (distinct from `wrv.value`) to hold the return value of `next`. That basically changes `hasNext` and `next` as follows and removes `advance`.

```scala
override def hasNext: Boolean = haveNext
private var rv: RegionValue = RegionValue()
override def next() = {
  rv.set(wrv.value.region, wrv.value.offset)
  while (bit.hasNext && ort.kInRowOrd.compare(wrv.value, bit.head) == 0)
    bit.next()
  haveNext = bit.hasNext
  if (haveNext)
    wrv.set(bit.next())
  rv
}
```

Reply:
Actually I don't think that works because when I call `bit.next` I will (possibly) reset the region that was captured in `rv`. I think I need a second `WritableRegionValue` to make this work, but now I'm holding two copies: the return-value and the next-value. These are in addition to the contents of `bit.head` which is potentially-the-next-next-value.

---

I'm open to other approaches, but I ran down a bunch of dead ends trying to avoid `stale` and couldn't swing it with only one copy of a region value.

Reply:
Actually a simpler solution came to me last night while falling asleep. I'll write it up later today.